subsys/fs/nvs/nvs.c

```diff
@@ -6,104 +6,492 @@
+ * maximum allowed len is 16383 byte.
+ */
+size_t _nvs_put_length(u8_t *buf, size_t len)
 {
-	return entry->data_addr + _nvs_len_in_flash(fs, entry->len);
+	len &= 0x3fff;
+	if (len < 0x80) {
+		buf[0] = len;
+		return sizeof(u8_t);
+	}
+	buf[0] = (len & 0x7f) | 0x80;
+	buf[1] = len >> 7;
+	return sizeof(u16_t);
 }
 
-int _nvs_bd_check(struct nvs_fs *fs, off_t offset, size_t len)
+/* end basic calculation routines */
+
+/* basic flash routines */
+
+/* basic flash write */
+int _nvs_flash_write(struct nvs_fs *fs, off_t offset, const void *data,
+		     size_t len)
 {
-	if ((offset > fs->sector_size * fs->sector_count) ||
-	    (offset + len > fs->sector_size * fs->sector_count)) {
-		/* operation outside nvs area */
-		return -EINVAL;
+	int rc;
+
+	rc = flash_write_protection_set(fs->flash_device, 0);
+	if (rc) {
+		/* flash protection set error */
+		return rc;
 	}
-	if ((offset & ~(fs->sector_size - 1)) !=
-	    ((offset + len - 1) & ~(fs->sector_size - 1))) {
-		/* operation over sector boundary */
-		return -EINVAL;
+	rc = flash_write(fs->flash_device, fs->offset + offset, data, len);
+	if (rc) {
+		return rc;
 	}
+
+	(void) flash_write_protection_set(fs->flash_device, 1);
 	return 0;
 }
 
-void _nvs_addr_advance(struct nvs_fs *fs, off_t *addr, u16_t step)
+/* basic flash read */
+int _nvs_flash_read(struct nvs_fs *fs, off_t offset, void *data,
+		     size_t len)
 {
-	*addr += step;
-	if (*addr >= (fs->sector_count * fs->sector_size)) {
-		*addr -= (fs->sector_count * fs->sector_size);
+	int rc;
+
+	rc = flash_read(fs->flash_device, fs->offset + offset, data, len);
+	if (rc) {
+		return rc;
 	}
+	return 0;
 }
 
-/* Getting the sector header for address given in offset */
-int _nvs_sector_hdr_get(struct nvs_fs *fs, off_t offset,
-			struct _nvs_sector_hdr *sector_hdr)
+/* basic flash erase */
+int _nvs_flash_erase(struct nvs_fs *fs, off_t offset, size_t len)
 {
-	return flash_read(fs->flash_device,
-			  fs->offset + (offset & ~(fs->sector_size - 1)),
-			  sector_hdr,
-			  sizeof(*sector_hdr));
+	int rc;
+
+	rc = flash_write_protection_set(fs->flash_device, 0);
+	if (rc) {
+		/* flash protection set error */
+		return rc;
+	}
+	SYS_LOG_DBG("Erasing flash at %"PRIx32", len %d", offset, len);
+	rc = flash_erase(fs->flash_device, fs->offset + offset, len);
+	if (rc) {
+		/* flash write error */
+		return rc;
+	}
+	(void) flash_write_protection_set(fs->flash_device, 1);
+	return 0;
 }
 
-/* Initializing sector by writing magic and status to sector header,
- * set fs->write_address just after header offset should be pointing
- * to the sector
+/* end of basic flash routines */
+
+/* advanced flash routines */
+
+/* _nvs_flash_block_cmp compares the data in flash at offset to data
+ * in blocks of size NVS_BLOCK_SIZE, returns 0 if equal
  */
-int _nvs_sector_init(struct nvs_fs *fs, off_t offset)
+int _nvs_flash_block_cmp(struct nvs_fs *fs, off_t offset, const void *data,
+		   size_t len)
+{
+	int rc, block_size;
```

Comment:
You should define block_size as size_t, as it's the type returned by sizeof and also the type of len, so that the comparison is done on the same type.

Reply:
Correct block_size should be size_t, will change